### ISSUE TITLE
kube-apiserver: --advertise-address/--bind-address not working with "blank" ip

### ISSUE THREAD

**Issue by klausenbusk** [kubernetes/kubernetes#58686]
**Is this a BUG REPORT or FEATURE REQUEST?**:

> /kind bug

**What happened**:

The help for `kube-apiserver` states the following:
```
      --advertise-address ip                                    The IP address on which to advertise the apiserver to members of the cluster. This address must be reachable by the rest of the cluster. If blank, the --bind-address will be used. If --bind-address is unspecified, the host's default interface will be used.
      --bind-address ip                                         The IP address on which to listen for the --secure-port port. The associated interface(s) must be reachable by the rest of the cluster, and by CLI/web clients. If blank, all interfaces will be used (0.0.0.0). (default 0.0.0.0)
```
but that doesen't seems to be the case.
Setting either `--advertise-address` or `--bind-address` blank, yields the following error:

```
root@84e925557f65:/# kube-apiserver --bind-address="" |& head -n 2
Error: invalid argument "" for --bind-address=: failed to parse IP: ""

root@84e925557f65:/# kube-apiserver --advertise-address="" |& head -n 2
Error: invalid argument "" for --advertise-address=: failed to parse IP: ""
```

**What you expected to happen**:

`kube-apiserver` should start as expected.

**How to reproduce it (as minimally and precisely as possible)**:

Try starting `kube-apiserver` with either `--advertise-address` or `--bind-address` blank.

**Anything else we need to know?**:

I stumbled on this behaviour, when trying to add the required change for supporting CCM (cloud controller manager) in bootkube (https://github.com/kubernetes-incubator/bootkube/pull/804, https://github.com/kubernetes-incubator/bootkube/pull/803). bootkube use `status.podIP` as the `--advertise-address` for the bootstrap-apiserver and the real apiserver, but when starting `kubelet` with `--cloud-provider=external` `status.podIP` is empty, and so `kube-apiserver` is started with `--advertise-address=` which the help page says is supported.

This should be fixed so we can start a self-hosted k8s cluster with a external CCM, without ending up in a chicken-egg situation.

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.2", GitCommit:"5fa2db2bd46ac79e5e00a4e6ed24191080aa463b", GitTreeState:"clean", BuildDate:"2018-01-18T10:09:24Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.2", GitCommit:"5fa2db2bd46ac79e5e00a4e6ed24191080aa463b", GitTreeState:"clean", BuildDate:"2018-01-18T09:42:01Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}
```

/sig api-machinery

**Comment by klausenbusk** [comment]
cc @kubernetes/sig-api-machinery-bugs

**Comment by dims** [comment]
i'd say we fix the comment :) 

```
If not specified, all interfaces will be used (0.0.0.0). (default 0.0.0.0)
```

**Comment by jennybuckley** [comment]
@cheftako

**Comment by j0gger** [comment]
Happy to volunteer

**Comment by colemickens** [comment]
@dims Are you saying that's what the comment should be updated to, or that part should be removed?

(This is from me reading through this for the first time, and actually demonstrated by the two cases the OP has posted above)

The `--advertise-address` does not, in any case, default to `--bind-address`. It also does not default to `0.0.0.0`. In fact, even if you do specify `0.0.0.0`, the code still chooses the default interface. https://github.com/kubernetes/apiserver/blob/master/pkg/server/options/server_run_options.go#L77 -> https://github.com/kubernetes/apimachinery/blob/2f1e02d3e57b8fb5206c5326bcb65217edc63a8e/pkg/util/net/interface.go#L383

I think `server_run_options.go` should be updated to set the default even when `--advertise-address=""` so that it works like other arguments, like `--cloud-provider=""`. (and the reference to inheriting `--bind-address` removed, or the code fixed so that it does that as well.) Probably `--bind-address=""` should get the same fix.

**Comment by dims** [comment]
@colemickens makes sense. thanks for digging in. (my first thought was that if there was a mismatch between the help text and functionality then we could have just fixed up the help text, but your analysis makes more sense!)

**Comment by j0gger** [comment]
@colemickens What should be the default? 0.0.0.0? Or should it retain the current behavior of picking up hosts default interface?

**Comment by colemickens** [comment]
@j0gger I don't know. I don't know if coercing `0.0.0.0` to `default interface` was intentional or not. Someone else will have to weigh in.

**Comment by j0gger** [comment]
@smarterclayton @colemickens I'm happy to take this on, but need some direction here. If not you, who can offer some direction here? I am fairly new to Kubernetes.

**Comment by terinjokes** [comment]
> If blank, all interfaces will be used (0.0.0.0 for all IPv4 interfaces and :: for all IPv6 interfaces).

Based on this description, I was expecting `--bind-address` to be the way to dual-stack the apiserver.

**Comment by v-ilinski** [comment]
This issue still exists. If kubeadm is used,  there's no way to get a kube-apiserver manifest without an "--advertise-address" argument. If it is set to an empty string in kubeadm.conf via the apiServer/extraArgs, the kube-apiserver maniest will have it set to an empty string also, which will prevent kube-apiserver from starting ...due to the error in the description above.

/reopen

**Comment by lavalamp** [comment]
I suspect the code is treating omitting the argument completely as different than passing it the empty string.

**Comment by duhruh** [comment]
@v-ilinski I can't seem to reproduce this on my machine with `kubeadm`, from a clean install I ran `kubeadm init --config config.yaml`

my relevant config:
```yaml
...
apiServer:
  timeoutForControlPlane: 4m0s
apiVersion: kubeadm.k8s.io/v1beta2
certificatesDir: /etc/kubernetes/pki
clusterName: kubernetes
controllerManager:
  extraArgs:
    advertiseAddress: ""
...
imageRepository: k8s.gcr.io
kind: ClusterConfiguration
kubernetesVersion: v1.20.1
...
```

note I tested this with  `localAPIEndpoint.advertiseAddress` not set, then again with it set to `""`

In either case my kube-apiserver runs successfully.

My setup:
```
Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.1", GitCommit:"c4d752765b3bbac2237bf87cf0b1c2e307844666", GitTreeState:"clean", BuildDate:"2020-12-18T12:09:25Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.1", GitCommit:"c4d752765b3bbac2237bf87cf0b1c2e307844666", GitTreeState:"clean", BuildDate:"2020-12-18T12:00:47Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
```


I could have missed something, can you still reproduce this? If so could you provide the steps?

**Comment by v-ilinski** [comment]
@duhruh the extraArgs/advertise-address should be under the apiServer, rather than controllerManager. (note advertise-address instead of advertiseAddress) 

```yaml
...
apiServer:
  timeoutForControlPlane: 4m0s
  extraArgs:
    advertise-address: ""
...
```

Would you be able to share the resulting kube-apiserver.yaml manifest? Mine looks like:

```yaml
...
spec:
  containers:
  - command:
    - kube-apiserver
    - --advertise-address=
...
```
